documentation/configuration.md
<a name="configuration"></a>
## Configuration
There are two ways of configuring Perseo CEP:
* The default Basic Configuration is read from the `config.js` file in the root of the project folder.
* Some pieces of configuration can be overriden using environment variables, as it is explained in the following section.
 
### Environment Variables Configuration
The following table shows the environment variables available for Perseo configuration:

| Environment variable      | Description                            |
|:------------------------- |:-------------------------------------- |
| PERSEO_ENDPOINT_HOST      | Host where the CEP will listen.        |
| PERSEO_ENDPOINT_PORT      | Port where the CEP will listen.        |
| PERSEO_MONGO_ENDPOINT     | Endpoint (host[:port]) list for Mongo DB.            |
| PERSEO_MONGO_REPLICASET   | ReplicaSet name for Mongo DB.          |
| PERSEO_MONGO_USER         | User for Mongo DB.                     |
| PERSEO_MONGO_PASSWORD     | Password for Mongo DB.                 g|
| PERSEO_CORE_URL           | Full URL where Perseo Core is listening (e.g: http://63.34.124.1:8080). |
| PERSEO_NEXT_URL           | Full URL where Perseo Core replicated node is listening. Same format as above. |
| PERSEO_ORION_URL          | Full URL of the Orion Context Broker (e.g: http://64.124.28.15:1026).          |
| PERSEO_LOG_LEVEL          | Log level.         |
| PERSEO_SMTP_HOST          | Host of the SMTP server |
| PERSEO_SMTP_PORT          | Port of the SMTP server |
| PERSEO_SMTP_VERIFY_CA     | `true` if self-signed or invalid TLS certificate should be rejected |
| PERSEO_SMTP_SECURE        | `true` if SSL should be used with the SMTP server |
| PERSEO_SMTP_AUTH_USER     | Authentication data, the username |
| PERSEO_SMTP_AUTH_PASS     | Authentication data, the password for the user |
| PERSEO_SMTP_TLS_REJECTUNAUTHORIZED     | Reject if unauthorized security is found (i.e. selfsigned certificates). Default is false. |
| PERSEO_SMS_URL            | URL for sending SMSs (SMPP Adapter) |
| PERSEO_SMS_API_KEY        | API KEY for sending SMSs, if necessary. Only for the SMPP Adapter simulator |
| PERSEO_SMS_API_SECRET     | API SECRET for sending SMSs, if necessary. Only for the SMPP Adapter simulator |
| PERSEO_SMS_FROM           | Field `from` for the outgoing SMSs. Required by the SMPP Adapter |
| PERSEO_SMPP_HOST          | Host of the SMPP server |
| PERSEO_SMPP_PORT          | Port of the SMPP server |
| PERSEO_SMPP_SYSTEMID      | SystemID for the user of the SMPP server |
| PERSEO_SMPP_PASSWORD      | Password for the user of the SMPP server |
| PERSEO_SMPP_FROM          | Number from SMS are sending by SMPP server |
| PERSEO_SMPP_ENABLED       | SMPP is default method for SMS instead of use SMS gateway |
| PERSEO_NOTICES_PATH       | Path for incoming notices, default value '/notices' |
| PERSEO_RULES_PATH         | Path for incoming rules, default value '/rules' |

### Basic Configuration
In order to have perseo running, there are several basic pieces of information to fill:
* `config.logLevel`: level for log messages (`FATAL`, `ERROR`, `INFO` or `DEBUG`)
* `config.perseoCore.rulesURL`: URL for management of EPL rules at core.
* `config.perseoCore.noticesURL`: URL for processing events at core rule engine.
* `config.perseoCore.interval`: interval for refreshing rules at core rule engine (milliseconds).
* `config.smtp.port`: port for sending email.
* `config.smtp.host`:  host for sending email.
* `config.smtp.secure`:  defines if the connection should use SSL (if true) or not (if false).
* `config.smtp.tls.rejectUnauthorized`:  defines if if self-signed or invalid TLS certificate should be rejected (if true) or not (if false).
* `config.smtp.auth.user`:  authentication data, the username.
* `config.smtp.auth.pass`:  authentication data, the password for the user.
* `config.smtp.tls.rejectUnauthorized`: reject if unauthorized security is found (i.e. selfsigned certificates). Default is false.
* `config.sms.URL`: URL for sending SMSs.
* `config.sms.from`: Field `from` for the outgoing SMSs. Required by the SMPP Adapter.
* `config.sms.API_KEY`: API KEY for sending SMSs, if necessary. Only for the SMPP Adapter simulator.
* `config.sms.API_SECRET`: API SECRET for sending SMSs, if necessary. Only for the SMPP Adapter simulator.
* `config.smpp.host`: Host of the SMPP server.
* `config.smpp.port`: Port of the SMPP server.
* `config.smpp.systemid`: SystemID for the user of the SMPP server
* `config.smpp.password`: Password for the user of the SMPP server
* `config.smpp.from`: Number from SMS are sending by SMPP server
* `config.smpp.enabled`: SMPP is default method for SMS instead of use SMS gateway.
* `config.orion.URL`: Context Broker base URL, e.g. https://orion.example.com:1026
* `config.mongo.URL`: URL for connecting mongoDB.
* `config.executionsTTL`: Time-To-Live for documents of action executions (seconds).
* `config.checkDB.delay`:  Number of milliseconds to check DB connection (see [database aspects](admin.md#database-aspects) documentation for mode detail).
* `config.checkDB.reconnectTries`:  Number of of attempts to reconnect (see [database aspects](admin.md#database-aspects) documentation for mode detail).
* `config.checkDB.reconnectInterval`:  Number of milliseconds to wait between attempts to reconnect (see [database aspects](admin.md#database-aspects) documentation for mode detail).
* `config.checkDB.bufferMaxEntries`:  Number of operations buffered up before giving up on getting a working connection (see [database aspects](admin.md#database-aspects) documentation for mode detail).



Options for HA:
* `config.isMaster`: `true` if this one is the master or `false` it it is the slave.
* `config.slaveDelay`: Slave's delay to try to execute an action (milliseconds).
* `config.nextCore.rulesURL`: URL for management of EPL rules at *replicated* core. If set, the rules will be propagated to that one also.
* `config.nextCore.noticesURL`: URL for processing events at *replicated* core rule engine. If set, the events will be propagated to that one also.

Options for Authentication through PEP (for update action)
* `config.authentication.host`: host (keyStone) to exchange trust tokens for access tokens
* `config.authentication.port`: port,
* `config.authentication.user`: provisioned user for CEP in Keystone
* `config.authentication.password`: provisioned password for CEP in Keystone


URL format for mongoDB could be found at http://mongodb.github.io/node-mongodb-native/driver-articles/mongoclient.html


config.js
/*
 * Copyright 2015 Telefonica Investigación y Desarrollo, S.A.U
 *
 * This file is part of perseo-fe
 *
 * perseo-fe is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * perseo-fe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with perseo-fe.
 * If not, see http://www.gnu.org/licenses/.
 *
 * For those usages not covered by the GNU Affero General Public License
 * please contact with::[contacto@tid.es]
 */

'use strict';

var config = {};

/**
 * Default log level. Can be one of: 'DEBUG', 'INFO', 'WARN', 'ERROR', 'FATAL'
 * @type {string}
 */
config.logLevel = 'INFO';


/**
 * Configures the exposed API.
 */
config.endpoint = {
    host: 'localhost',
    port: 9090,
    rulesPath : '/rules',
    actionsPath : '/actions/do',
    noticesPath : '/notices',
    vrPath : '/m2m/vrules',
    checkPath : '/check',
    versionPath : '/version',
    logPath: '/admin/log',
    metricsPath: '/admin/metrics'
};

/**
 * Is Master this one?
 */
config.isMaster = true;

/**
 * Delay (in milliseconds) for slave to execute an action
 */
config.slaveDelay = 500;


/**
 * DB Configuration.
 */
config.mongo = {
    // The URI to use for the database connection. It supports replica set URIs.
    // mongodb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]
    // I.e.: 'mongodb://user:pass@host1:27017,host2:27018,host3:27019/cep?replicaSet=myrep'
    url : 'mongodb://localhost:27017/cep'
};

/**
 * OrionDB Configuration.
 */
config.orionDb = {
    url : 'mongodb://localhost:27017/orion',
    collection : 'entities',
    prefix : 'orion',
    batchSize: 500
};


/**
 * EPL core options
 *
 * interval is the time in milliseconds between refreshing rules
 * at core. Each <<interval>> ms, the rules are sent to core.
 */
config.perseoCore = {
    rulesURL : 'http://localhost:8080/perseo-core/rules',
    noticesURL : 'http://localhost:8080/perseo-core/events',
    interval: 60e3*5
};
/**
 * NEXT EPL core options (with HA)
 */
config.nextCore = {
    //
    // Note this parameter is empty, so Perseo will not use HA by default
    //
    /*
    rulesURL : 'http://next-core:8080/perseo-core/rules',
    noticesURL : 'http://next-core:8080/perseo-core/events'
    */
};


/**
 * SMTP endpoint options
 */
config.smtp = {
    port: 25,
    host: 'smtpserver',
    secure: false,
    tls: {
        rejectUnauthorized: true
    }
    /*
     ,
     auth: {
     user: 'abc',
     pass: 'xyz'
     }
    */
    ,
    tls: {
      // keep rejectUnauthorized to false when secure is also false
      rejectUnauthorized: false
    }
    
};

/**
 * SMS endpoint options
 */
config.sms = {
    URL : 'http://sms-endpoint/smsoutbound',
    API_KEY : '',
    API_SECRET: '',
    from: 'tel:22012;phone-context=+34'
};

/**
 * SMPP endpoint options
 */
config.smpp = {
    host: '',
    port: '',
    systemid: '',
    password: '',
    from: '346666666',
    enabled: false
};

/**
 * Orion (Context Broker) endpoint options
 */
config.orion = {
        URL : 'http://orion-endpoint:1026/NGSI10/updateContext'
};

/**
 * Authorization endpoint
 */
config.authentication = {
    host: 'keystone',
    port: '5001',
    user: 'user',
    password: 'password'
};

/**
 * Collections
 * @type {{}}
 */
config.collections = {
    rules : 'rules',
    executions: 'executions'
};

/**
 * Executions TTL
 *
 * Number of seconds to expire a document in 'executions'
 * @type {Number}
 *
 */
config.executionsTTL = 1 * 24 * 60 * 60;

/**
 * Constants for missing header fields for service (Fiware-servicepath) DEFAULT_SUBSERVICE
 * and tenant (Fiware-service) DEFAULT_TENANT
 *
 * @type {{}}
 */
config.DEFAULT_SUBSERVICE= '/';
config.DEFAULT_SERVICE= 'unknownt';

/**
 * CheckDB configuration
 *
 * delay Number of milliseconds to check DB connection
 * @type {Number}
 *
 * reportInterval Number of milliseconds to report a problem with DB connection in logs
 * @type {Number}
 */
config.checkDB = {
    delay: 5e3,
    reconnectTries: 1e3,
    reconnectInterval: 5e3,
    bufferMaxEntries: 5
};

/**
 * Rest base for changing access to twitter
 *
 * It should not be changed in normal conditions
 *
 * restBase:  URL
 * @type {String}
 *
 */
config.restBase = null;

module.exports = config;
